web/src/desktop_notifications.ts

```diff
@@ -0,0 +1,112 @@
+import $ from "jquery";
+
+import type {NotificationData} from "./electron_bridge";
+import type {Message} from "./message_store";
+
+export type NoticeMemEntry = {
+    obj: ElectronBridgeNotification | Notification;
+    msg_count: number;
+    message_id: number;
+};
+
+export const notice_memory: Map<string, NoticeMemEntry> = new Map();
+
+export let NotificationAPI: typeof ElectronBridgeNotification | typeof window.Notification;
+
+export function set_notification_api(
+    n: typeof ElectronBridgeNotification | typeof window.Notification,
+): void {
+    NotificationAPI = n;
+}
+
+class ElectronBridgeNotification extends EventTarget {
+    title: string = "";
+    dir: NotificationDirection = "auto";
+    lang: string = "";
+    body: string = "";
+    tag: string = "";
+    icon: string = "";
+    data: unknown;
+    close: () => void = (): void => {};
+    constructor(title: string, options: NotificationOptions) {
+        super();
+        let notification_data: NotificationData = {
+            title: "",
+            dir: "auto",
+            lang: "",
+            body: "",
+            tag: "",
+            icon: "",
+            data: "",
+            close: () => {},
+        };
+        if (window.electron_bridge && window.electron_bridge.new_notification) {
```

Comment:
Just throw an error if this invariant is violated. Don’t create a bunch of phony default values that will go and confuse other code down the line.

Reply:
I am sorry but I did not understand what are you trying to say by throwing an error if invariant is violated.